pkg/storage/streams/store.go

```diff
@@ -109,157 +111,203 @@ func (s *streamStore) Put(ctx context.Context, path storj.Path, pathCipher storj
 		default:
 		}
 	}()
 
 	derivedKey, err := encryption.DeriveContentKey(path, s.rootKey)
 	if err != nil {
-		return Meta{}, currentSegment, err
+		return Meta{}, currentSegmentIndex, err
+	}
+
+	encPath, err := EncryptAfterBucket(ctx, path, pathCipher, s.rootKey)
+	if err != nil {
+		return Meta{}, currentSegmentIndex, err
 	}
 
 	eofReader := NewEOFReader(data)
 
 	for !eofReader.isEOF() && !eofReader.hasError() {
-		// generate random key for encrypting the segment's content
-		var contentKey storj.Key
-		_, err = rand.Read(contentKey[:])
-		if err != nil {
-			return Meta{}, currentSegment, err
-		}
-
-		// Initialize the content nonce with the segment's index incremented by 1.
-		// The increment by 1 is to avoid nonce reuse with the metadata encryption,
-		// which is encrypted with the zero nonce.
-		var contentNonce storj.Nonce
-		_, err := encryption.Increment(&contentNonce, currentSegment+1)
-		if err != nil {
-			return Meta{}, currentSegment, err
-		}
-
-		encrypter, err := encryption.NewEncrypter(s.cipher, &contentKey, &contentNonce, s.encBlockSize)
-		if err != nil {
-			return Meta{}, currentSegment, err
-		}
-
-		// generate random nonce for encrypting the content key
-		var keyNonce storj.Nonce
-		_, err = rand.Read(keyNonce[:])
-		if err != nil {
-			return Meta{}, currentSegment, err
-		}
 
-		encryptedKey, err := encryption.EncryptKey(&contentKey, s.cipher, derivedKey, &keyNonce)
+		segmentEncrypter, err := newSegmentEncrypter(derivedKey, s.cipher, currentSegmentIndex, s.encBlockSize)
 		if err != nil {
-			return Meta{}, currentSegment, err
+			return Meta{}, currentSegmentIndex, err
 		}
 
 		sizeReader := NewSizeReader(eofReader)
 		segmentReader := io.LimitReader(sizeReader, s.segmentSize)
 		peekReader := segments.NewPeekThresholdReader(segmentReader)
-		largeData, err := peekReader.IsLargerThan(encrypter.InBlockSize())
+
+		// If the data is larger than the inline threshold size, then it will be a remote segment
+		isRemote, err := peekReader.IsLargerThan(s.inlineThreshold)
 		if err != nil {
-			return Meta{}, currentSegment, err
-		}
-		var transformedReader io.Reader
-		if largeData {
-			paddedReader := eestream.PadReader(ioutil.NopCloser(peekReader), encrypter.InBlockSize())
-			transformedReader = encryption.TransformReader(paddedReader, encrypter, 0)
-		} else {
-			data, err := ioutil.ReadAll(peekReader)
-			if err != nil {
-				return Meta{}, currentSegment, err
-			}
-			cipherData, err := encryption.Encrypt(data, s.cipher, &contentKey, &contentNonce)
-			if err != nil {
-				return Meta{}, currentSegment, err
-			}
-			transformedReader = bytes.NewReader(cipherData)
+			return Meta{}, currentSegmentIndex, err
 		}
 
-		putMeta, err = s.segments.Put(ctx, transformedReader, expiration, func() (storj.Path, []byte, error) {
-			encPath, err := EncryptAfterBucket(ctx, path, pathCipher, s.rootKey)
+		// Deal with inline first
+		if !isRemote {
+			data := peekReader.Buf()
```

Comment:
I think it might make sense to use `data, err := ioutil.ReadAll(peekReader)` rather than adding a `Buf()` function to `peekReader`

Alternatively, if you still want to go with the `Buf()` route, maybe there should be a check within that function to verify that `IsLargerThan` was called? Otherwise there is the risk of returning an empty buffer if someone forgot to call `IsLargerThan` first

Reply:
yeah good point, i think we can just `readAll`.  I'm trying to remember if there is a reason we didn't do that in the first place. Either way, yes to adding either `readAll` or adding the check to `IsLargerThan`